src/util/thread.cpp

```diff
@@ -10,10 +10,11 @@
 
 #include <exception>
 #include <functional>
+#include <string>
 
-void util::TraceThread(const char* thread_name, std::function<void()> thread_func)
+void util::TraceThread(std::string_view thread_name, std::function<void()> thread_func)
```

Comment:
I agree, thanks, good catch. Made it a `const std::string` instead, and since we're already copying here I re-reverted the logic for `ThreadRename()` to use an lvalue ref again.

Reply:
Do you mean for `util::TraceThread()`? I think we explicitly want a copy here to not have to worry about lifetime? The copying I mentioned was further downstream for when we pass `thread_name` to util::ThreadRename()`, where indeed a `const std::string&` is used to avoid copying again.

Reply:
`TraceThread` does not return before the end of the thread, so the reference is implicitly stored as a bound argument for the whole duration of the thread, no?

Reply:
> You probably want a `const std::string&` to avoid a copy.

.

Reply:
@MarcoFalke if you agree I'll revert to `string_view` again, I'm not feeling very confident here but what sipa says makes sense to me. At the very least, it shouldn't introduce any _new_ lifetime issues compared to what we had before with `const char*`?

Reply:
Oh, I see. The arguments to `std::thread` are copied, unless they are passed as `std::ref` ( https://en.cppreference.com/w/cpp/thread/thread/thread )

So `const std::string&` is indeed fine here, as it is a reference to the copy that was created by the `std::thread` constructor. (However, `std::string` should be fine as well and come with only the overhead equivalent to constructing an empty `std::string`)

Reply:
For reference, the following gives an `AddressSanitizer: stack-use-after-scope`, because the argument was passed with `std::ref`, but does not if the argument is not passed with `std::ref`.

<details><summary>`std::thread`/`std::ref` test</summary>

```cpp
#include <chrono>
#include <iostream>
#include <thread>

void fun(const std::string &a) {
  std::this_thread::sleep_for(std::chrono::milliseconds{10});
  std::cout << a << std::endl;
}

int main() {
  std::thread a;
  {
    std::string s{"s"};
    a = std::thread{fun, std::ref(s)};
  }
  a.join();
}
```

</details>

Reply:
To clarify, anything is fine here, but `string_view` can lead to lifetime issues if there is a leftover caller that uses c_str on a temporary string. For example, this is UB:


```cpp
#include <chrono>
#include <iostream>
#include <thread>

void fun(std::string_view a) {
  std::this_thread::sleep_for(std::chrono::milliseconds{10});
  std::cout << a << std::endl;
}

int main() {
  std::thread a;
  {
    std::string s{"s"};
    a = std::thread{fun, s.c_str()};
  }
  a.join();
}

Reply:
Thanks for the background info MarcoFalke, that's really helpful.

> can lead to lifetime issues if there is a leftover caller that uses c_str on a temporary string

Would you agree that that would just be bad implementation, and not something we should expect downstream functions to try and fix?

Reply:
Previously passing a `std::string` was only possible by calling `c_str()` on it, so this UB may persist in a very unlikely and unfortunate silent merge conflict. (Changing `const char*` to `std::string_view` shouldn't introduce any new UB, but it wouldn't fix it either, if it existed previously). Though, again that seems unlikely so I think anything is fine here and this is almost a style question.

Reply:
FWIW, in our code base, all `util::TraceThread` callers pass either a string literal (in most cases) or a `BaseIndex::GetName()` return value of `const std::string&` type (the only case).